src/Interpreters/RedundantFunctionsInOrderByVisitor.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+
+namespace DB
+{
+
+class RedundantFunctionsInOrderByMatcher
```

Comment:
Add a commnet that you optimize `ORDER BY x, y, f(x), g(x, y), f(h(x))` rewriting it into `ORDER BY x, y` in case if `f()`, `g()`, `h()` are deterministic (in scope of query).

It also valuable to note that you do not optimize cases `ORDER BY x, h(f(x), g(x))` and `ORDER BY f(x), g(x), x` even if `f(x)` is bijection for `x` or `g(x)`.